### PR TITLE
Migrate Unique User ID from Registry to Filesystem

### PR DESCRIPTION
**Pull request by epresi**

### CODE REVIEW

File: TechTalk.SpecFlow.VsIntegration.Implementation/Analytics/FileUserIdStore.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        public FileUserIdStore(IWindowsRegistry windowsRegistry, IFileService fileService, IDirectoryService directoryService)
+        {
+            _windowsRegistry = windowsRegistry;
+            _fileService = fileService;
+            _directoryService = directoryService;
+            _lazyUniqueUserId = new Lazy<string>(FetchAndPersistUserId);
+        }
+
+        public string Get()
```
Comment (SabotageAndi) [inline TechTalk.SpecFlow.VsIntegration.Implementation/Analytics/FileUserIdStore.cs]: I would rename this method to `GetUserId`.

File: TechTalk.SpecFlow.VsIntegration.Implementation/Services/DirectoryService.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+        public void CreateDirectory(string path)
+        {
+            Directory.CreateDirectory(path);
+
+        }
+
+        public string GetDirectoryName(string path)
+        {
+            return Path.GetDirectoryName(path);
```
Comment (SabotageAndi) [inline TechTalk.SpecFlow.VsIntegration.Implementation/Services/DirectoryService.cs]: I am always unsure if abstracting the `Path.*` methods is a good idea. They do string manipulation and don't have side effects.

File: TechTalk.SpecFlow.VsIntegration.Implementation/Services/WindowsRegistry.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+namespace TechTalk.SpecFlow.VsIntegration.Implementation.Services
+{
+    public class WindowsRegistry : IWindowsRegistry
+    {
+        public object GetValueForCurrentUser(string registryPath, string registryValueName, object defaultValue)
+        {
+            var rootKey = Registry.CurrentUser;
+            var key = rootKey.OpenSubKey(registryPath);
+
+            if (key == null)
```
Comment (SabotageAndi) [inline TechTalk.SpecFlow.VsIntegration.Implementation/Services/WindowsRegistry.cs]: `if (key is null)` is the current best practice for null checks.
Here the equality operator is not called.

File: UnitTests/VsIntegration.Implementation.UnitTests/FileUserIdStoreTests.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                    windowsRegistry.GetValueForCurrentUser(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<object>()))
+                .Returns(userIdString);
+        }
+
+        private void GivenUserIdStringInFile(string userIdString)
+        {
+            fileServiceStub.Setup(fileService => fileService.ReadAllText(It.IsAny<string>())).Returns(userIdString);
+        }
+
+        private void GivenFileExists(bool fileExists)
```
Comment (nemesv) [inline UnitTests/VsIntegration.Implementation.UnitTests/FileUserIdStoreTests.cs]: Consider creating a separate method for the fileExists: false case for better readablility, e.g GivenFileDoesNotExists.

File: UnitTests/VsIntegration.Implementation.UnitTests/FileUserIdStoreTests.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        {
+            fileServiceStub.Setup(fileService => fileService.Exists(It.IsAny<string>())).Returns(fileExists);
+        }
+
+        private void GivenDirectoryExists(bool directoryExists)
```
Comment (nemesv) [inline UnitTests/VsIntegration.Implementation.UnitTests/FileUserIdStoreTests.cs]: Consider creating a separate method for the directoryExists: false case for better readablility, e.g GivenDirectoryDoesNotExists.

File: UnitTests/VsIntegration.Implementation.UnitTests/FileUserIdStoreTests.cs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+        [Test]
+        public void Should_GetUserIdFromFile_WhenFileExists()
+        {
+            GivenFileExists(true);
+            GivenUserIdStringInFile(UserId);
+
+            string userId = sut.Get();
+
+            fileServiceStub.Verify(fileService => fileService.ReadAllText(FileUserIdStore.UserIdFilePath));
```
Comment (nemesv) [inline UnitTests/VsIntegration.Implementation.UnitTests/FileUserIdStoreTests.cs]: In general the tests should not care about the internals of the implementation. It should be enough to assert on the end result: the correct guid was returned which was provided in the setup with the Givens. 

Calling `Verify` for every call couples the tests to the actual logic and any refactoring of the code would require to change also the tests.

File: TechTalk.SpecFlow.VsIntegration.Implementation/Analytics/FileUserIdStore.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+            return GenerateAndPersistUserId();
+        }
+
+        private void PersistUserId(string userId)
+        {
+            var directoryName = Path.GetDirectoryName(UserIdFilePath);
+            if (!Directory.Exists(directoryName))
+            {
+                Directory.CreateDirectory(directoryName);
```
Comment (SabotageAndi) [inline TechTalk.SpecFlow.VsIntegration.Implementation/Analytics/FileUserIdStore.cs]: Don't we now create directories in the test run?

**Comment by SabotageAndi** [comment]
I am fine with these changes now.
@nemesv what about you?

**Comment by nemesv** [comment]
LGTM 😄

**Pull request merged**